feat(formComponent): add Form, FormGroup, ActionGroup PF4 components

rvsia opened the pull request:
**What**:

Adds Form, FormGroup, ActionGroup component
issue: #390 

**Link to Storybook**:

http://chivalrous-scale.surge.sh/components/form

**Additional issues**:

There are also issues with testing of console.errors. In some cases they work, in others not. I skipped them.

karelhala commented:
Ping @Hyperkid123

dgutride commented on packages/patternfly-4/react-core/package.json:
```diff
@@ -46,10 +46,10 @@
   },
   "devDependencies": {
-    "@patternfly/patternfly-next": "1.0.44",
+    "@patternfly/patternfly-next": "1.0.46",
```
This number needs to match throughout all of the packages (or the build will fail, as it is now).

rvsia replied:
Fixed :)

tlabaj commented on packages/patternfly-4/react-core/src/components/Form/FormGroup.js:
```diff
@@ -0,0 +1,121 @@
+  const classesLabel = css(styles.formLabel);
+
+  if (typeof label === 'string') {
+    if (required === true) {
```
I am wondering if this is necessary.  Is there a use case where someone would want to have the label be something other than a string?  This label is also used for accessibility purposes, if it is a node, I don't know how that will work.

karelhala replied:
That is actually desired, given that you may have something really complex as part of Form value you also want to have something complex for label. For instance autogenerated label (thus function) or if someone would want to have for instance icon as a label (can't think of a great reason why, but we shouldn't force consumers one way or another).
It's somewhat of a standard to provide consumer as much freedom of a choice as possible. Of course provide some default features but allow some complex usecases as well.

Hyperkid123 replied:
@tlabaj This has become a standard approach in component libraries in React. The most common reasons for this are:
- Rendering icons instead of strings
- Rendering images
- Dynamically changing label value without re-rendering whole form
- Complete styling customization for non-standard use cases.

tlabaj replied:
I am not disagreeing with the implementation.  It just differs from what I see in core.  My interpretation of what I see in core is that they expect the label to be a string. I think we need to at least have a discussion with design and css before diverging from what is in core.  :)

tlabaj commented on packages/patternfly-4/react-core/src/components/Form/FormGroup.js:
```diff
@@ -0,0 +1,121 @@
+  invalid: false,
+  helpertext: undefined,
+  helpertextInvalid: undefined,
+  fieldId: undefined
```
the id should be required like it is in core.

karelhala replied:
Well, not quite all the time. The id should be required if label is text, for other use cases it's not really required. However you are correct that it should be required if user wants to use text as label, so function should be used here.
```JSX
const propTypes = {
  ...
  id: (props) => {
    if (props.label && typeof props.label === 'string' && !props.id) {
      return new Error('FormGroup requires ID if label is string.');
    }
  }
}
```

rvsia replied:
I added the function!

jgiardino replied:
This question might be more related to [previous comments](https://github.com/patternfly/patternfly-react/pull/667#pullrequestreview-159609445) that are now outdated about allowing the label to accept values other than string... But is it possible to catch and warn the consumer about possible accessibility issues? For example, if something other than a string is passed for the label, can we warn the consumer that they should check that text is included somewhere inside the `<label>` (because without providing an accessible form label that the form will fail accessibility checks)?

> The id should be required if label is text, for other use cases it's not really required. 

If the consumer does provide a label that is not text, then the input should have an `aria-label` attribute that includes the text label for the form field. Can we catch this case and warn the consumer that one of the following must be true:

1. `<label>` includes a text string and the `<input>` includes an `id`
    _or_
2. `<input>` includes an `aria-label`

rvsia replied:
Good idea! 

FormGroup throws an error now if:

1. **label is not a string**

- Children is an **array** and no child has an` aria-label` (at least one of the children has to have an `aria-label`)
- Child does not have an `aria-label`

2. **label is a string**
- Children is an **array** and no child has an` id` (at least one of the children has to have an `id`)
- Child does not have an `id`

tlabaj replied:
Great Updates! 
They only concern I have is  that you are only warning that an aria-label/id is needed if none of the children has one.  In order for the component to be accessible, all the visible children should have an aria label if the Label is not a string.  And if the label is a string, all the visible children should have an id.

jgiardino replied:
To make sure I understand the FormGroup, it sounds like it can have one label and more than one input element?

Each input element should have a unique text string associated with it. And a text string can be associated with an input element using the `for` attribute on a `<label>` or the `aria-label` attribute on the input itself. 

I agree with @tlabaj's comment above about catching potential a11y issues and warning the user. I just wanted to expand on this statement: 

> And if the label is a string, all the visible children should have an id.

A single label can only be associated with a single input element using the `for`/`id` relationship. Therefore, I think the correct logic would be:

- If the label is a string, and only one visible child in the array, then that child must have an id
- If the label is a string, and more than one visible child in the array, then all children must have aria-label with the exception of the first child. The first child does not need `aria-label` if the label `for` attribute references the `id` value of that child. 
  - _It could be argued that we could allow any one child to be referenced by the label and not limit it to the first child. It would just be weird from the user experience perspective._

rvsia replied:
It has not to be only input elements. For example, I can imagine to put there an input with some image. And in cases like that you don’t need id or aria-label on that image:

```
<FormGroup label='How would you name this cat?' fieldId='cat-id' >
     <img src=‘imageofawesomecat.png' />
     <TextInput id='cat-id’ />
</ FormGroup>
```

In ideal approach FormGroup is just styling wrapper around some input field (select, textInput, etc.) with support of the helperText and label. Nothing more.

And if not, I would let consumers to add more labels, if they want more inputs. You can use the component like that (however I think it does not make sense to use it with more inputs):

```
<FormGroup label='How would you name this cat?' fieldId='cat-id' >
     <img src=‘imageofawesomecat.png' />
     <TextInput id='cat-id’ />
     <Label for='cat2'>And any last name?</ Label>
     <TextInput id='cat-last-name' />
</ FormGroup>
```
There will be probably some (and big) problems with styling, and consumers should manage it themselves. But it's up to them. I think it is more about to give consumers more freedom how to use the component if they want to. But I'm not a designer or architect, so I don't know how strict we want to be or how much freedom we want to allow. And of course, everything of this should be mentioned in a consumer documentation.

jgiardino replied:
Thanks for the input (no pun intended), @rvsia and @Hyperkid123! The best suggestion I have at this time is to move forward with this PR, and then later revisit this in the context of a demo or sample app, and just test what the consumer experience is for common use cases. I think that will help us understand if there are any gaps in how we inform consumers, and also how to best address them.

tlabaj commented on packages/patternfly-4/react-core/src/components/Form/FormGroup.js:
```diff
@@ -0,0 +1,121 @@
+const renderHelperText = ({ text, classes, fieldId }) => {
+  if (typeof text === 'string') {
+    return (
+      <p className={classes} id={`${String(fieldId)}-helper`} aria-live="polite">
```
This assumes that he consumer will use this for the aria-describedby attribute.  You should apply the aria-described by attribute with the same id you use here, when you render the child.

karelhala replied:
Well we should give users some kind of freedom to mess with their forms. I understand that it's kinda overcomplicated to pass multiple times same ID, but given that this is just simplification of combination of label and input we should stick to same logic as when using general `label` and `input` where you have to specify the ID twice as well.

martinpovolny commented:
Good job, @rvsia !

rvsia commented on packages/patternfly-4/react-core/src/components/Form/FormGroup.js:
```diff
@@ -0,0 +1,121 @@
+        <label className={classesLabel} htmlFor={fieldId}>
+          {label}
+          <span className={classesRequired} aria-hidden="true">
+            &#42;
```
This a standard way to write an asterisk in HTML, please see https://ascii.cl/htmlcodes.htm

tlabaj replied:
I understand that. In my opinion I think it should be a constant though. It is good coding practice and makes the code more readable.

rvsia commented on packages/patternfly-4/react-core/src/components/Form/ActionGroup.js:
```diff
@@ -0,0 +1,28 @@
+import styles from '@patternfly/patternfly-next/components/Form/styles.css';
+
+const propTypes = {
+  children: PropTypes.node,
```
added :)

rvsia commented on packages/patternfly-4/react-core/src/components/Form/Form.js:
```diff
@@ -0,0 +1,37 @@
+import { FormContext } from './FormContext';
+
+const propTypes = {
+  children: PropTypes.node,
```
Added :)

matthewcarleton commented:
Hey Richard, Thanks for implementing this! I'd love to take a look at the storybook link, is that possible?

rvsia commented:
> Hey Richard, Thanks for implementing this! I'd love to take a look at the storybook link, is that possible?

Hello Matthew, I would like to deploy the storybook. However the automatic deployment isn't working here and I don't know why and how to start it. Can you please help me?

LHinson commented:
@dgutride can you help with the storybook issue?

patternfly-build commented:
PatternFly documentation deployment: https://667-pr-patternfly-react-patternfly.surge.sh

jschuler commented on package.json:
```diff
@@ -33,7 +33,7 @@
     "enzyme": "^3.3.0",
-    "enzyme-adapter-react-16": "^1.1.1",
+    "enzyme-adapter-react-16": "^1.5.0",
     "enzyme-context-patch": "^0.0.9",
```
I had added the context patch before to work with the previous version of enzyme adapter react. But looks like the latest version includes it? In that case we can remove the patch again.

jschuler replied:
This is what also is probably breaking the build since the context patch is trying to patch enzyme adapter

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

rvsia commented:
@jschuler 

Do you please know how to change props in the docs props table:

from `union: , ,` 
to  `string, node, object` 

according to `PropTypes.oneOfType([PropTypes.string, PropTypes.node, PropTypes.func])`?

:thinking: 

Thanks!

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

mcarrano commented:
I am not sure exactly what the examples are intended to convey, but they do not conform with form layout practice as intended for PatternFly 4. Can we model what was implemented  as a form example in the Core repo? There are issues with spacing and labeling of checkboxes. Compare with https://pf-next.netlify.com/demos/BasicForms/examples/

rvsia commented:
@mcarrano The examples had been made before I knew about the BasicForms demo. 

I updated it!

There are still problems with checkboxes and radios. I will work on them in different PRs (I opened checkbox PR yesterday) and then I will update them here.

rvsia commented on packages/patternfly-4/react-core/src/components/Form/FormGroup.js:
```diff
@@ -0,0 +1,168 @@
+  /** Helper text after the field when the field is isValid. It can be a simple text, a function or an object. */
+  helperTextInvalid: PropTypes.oneOfType([PropTypes.string, PropTypes.node, PropTypes.func]),
+  /** ID of the included field. It has to be the same for proper working. */
+  fieldId: props => {
```
Initially I named this 'id' but it's not an id of the formGroup, but id of the some input field (used in 'for' in label). It would be a little bit confusing. However I am open to change 'field' to something else.

rvsia commented:
@mcarrano Spacing should be fixed in new examples. In the old ones, the spacing was wrong because I didn't use the toolbarItem component.

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

karelhala commented on packages/patternfly-4/react-core/src/components/Form/FormGroup.js:
```diff
@@ -0,0 +1,168 @@
+  } else if (typeof label === 'object') {
+    return label;
+  }
+  if (isRequired) {
```
Optional: can you use inline condition? I think it's much easier to read such code and you don't have to use consts for classes.
```JSX
return (
  <label className={css(styles.formLabel)} htmlFor={fieldId}>
    {label}
    {isRequired && <span className={css(styles.formLabelRequired)} aria-hidden="true">
      {ASTERISK}
    </span>}
  </label>
)
```

karelhala commented on packages/patternfly-4/react-core/src/components/Form/FormGroup.js:
```diff
@@ -0,0 +1,168 @@
+const renderHorizontal = children => {
+  const classesHorizontal = css(styles.formHorizontalGroup);
+
+  return <div className={classesHorizontal}>{children}</div>;
```
Optional: I prefer inline function calls over variable assigns (kinda better for memory work in JS)
```JSX
const renderHorizontal = children => (
  <div className={css(styles.formHorizontalGroup)}>{children}</div>
);
```

karelhala commented on packages/patternfly-4/react-core/src/components/Form/FormGroup.js:
```diff
@@ -0,0 +1,168 @@
+  fieldId,
+  ...props
+}) => {
+  const classes = css(styles.formGroup, getModifier(styles, isInline && 'inline'), className);
```
Similiar to this constant

karelhala commented on packages/patternfly-4/react-core/src/components/Form/index.d.ts:
```diff
@@ -0,0 +1,3 @@
+export { default as Form, FormProps } from './Form';
```
Probably missed to add `export * from './Form';` into `packages/patternfly-4/react-core/src/components/index.d.ts` as well.

rvsia replied:
added!

rvsia commented:
@mcarrano I updated examples http://chivalrous-scale.surge.sh/components/form

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

rvsia commented on packages/patternfly-4/react-core/src/components/Form/Form.test.js:
```diff
@@ -0,0 +1,16 @@
+import toJson from 'enzyme-to-json';
+import Form from './Form';
+
+describe('Form component', () => {
```
I think it's a good practice to group all tests of one components (controllers, etc.) together under one describe. At least it's the way I used to write tests.

And If you use describes this way, you get better testing messages:

![screenshot from 2018-10-11 10-07-38](https://user-images.githubusercontent.com/32869456/46789967-1ba0f580-cd3e-11e8-98b8-683ad2e0bc5d.png)

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

amarie401 commented on packages/patternfly-4/react-core/src/components/Form/FormGroup.js:
```diff
@@ -0,0 +1,168 @@
+  /** Additional classes added to the FormGroup. */
+  className: PropTypes.string,
+  /** Label text before the field. */
+  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node, PropTypes.func]),
```
Function support won't be necessary here

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

dmiller9911 commented on packages/patternfly-4/react-core/src/components/Form/FormGroup.js:
```diff
@@ -0,0 +1,134 @@
+    {({ isHorizontal }) => (
+      <div {...props} className={css(styles.formGroup, getModifier(styles, isInline && 'inline'), className)}>
+        {React.isValidElement(label)
+          ? label
```
Same comment here as the Checkbox PR. I think the element should be rendered inside the label tag.
See: https://github.com/patternfly/patternfly-react/pull/717#discussion_r224796410

dmiller9911 commented on packages/patternfly-4/react-core/src/components/Form/FormGroup.js:
```diff
@@ -0,0 +1,134 @@
+
+const renderHelperText = ({ helperText, classes, fieldId }) => {
+  if (typeof helperText === 'object') {
+    return helperText;
```
Same here. Elements should be inside the helper text block so they are styled. Since this will remove the conditional it would be better to simply inline in the main render.

dmiller9911 commented on packages/patternfly-4/react-core/src/components/Form/FormGroup.js:
```diff
@@ -0,0 +1,134 @@
+              </label>
+            )}
+        {isHorizontal ? <div className={css(styles.formHorizontalGroup)}>{children}</div> : children}
+        {helperText &&
```
once the helper text is simplified this can be written as
```jsx
<p 
  className={css(styles.formHelperText, isValid ? styles.modifiers.error : styles.modifiers.info)} 
  id={`${fieldId}-helper`} 
  aria-live="polite"
>
   {isInvalid ? helperTextInvalid ? helperText}
</p>

rvsia commented:
@dmiller9911 Because of wrapping nodes into a helperText I had to change `<p>` element to `<div>`. (To allow to put there everything.) Do you think it can cause any problems?

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

jschuler commented:
Is the `pf-l-grid` parent container needed in the examples?
Also, can you make one of the form inputs in the example invalid?

Comparing the DOM with with (pf-next)[https://pf-next.com/demos/BasicForms/examples/] i also see some differences
- The text inputs are missing the name field
- I also did not see aria-errormessage being added to it in the code if aria-valid is false
- For the radio buttons and checkboxes in the examples, the class on the inputs should be `pf-c-check__input` instead of `pf-c-check`, and the `pf-c-check` class should be on a container per input/label pair

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh

rvsia commented:
@jschuler 
- I removed the grid
- I added examples with invalid fields
- I also added state values for inputs to allow to write in fields :sweat_smile: 
- `Name` attributes was added
- I think missing `aria-errormesssage` should be managed by TextInput, etc. components and it should be different issue than this
- Checkboxes and Radios: Yeah, they were just placeholders. In the newest update I used recently merged Checkbox component (it wasn't Surged, I updated Surge right now! I am sorry!) and Radio component is in progress (https://github.com/patternfly/patternfly-react/pull/756), so after merging I will update it here, but I think it's nothing important in this examples right now . However, I modified radios to look as in demo, so it should be OK.

patternfly-build commented:
PatternFly-React preview: https://667-pr-patternfly-react-patternfly.surge.sh
